### ISSUE TITLE
groupby with datetime DataArray fails with `AttributeError`

### ISSUE THREAD

**Issue by j08lue** [pydata/xarray#1132]
I want to group some data by Oct-May season of each year, i.e. [(Oct 2000 - May 2001), (Oct 2001 - May 2002), ...]. I.e. I do not want some DJF-like mean over all the data but one value for each year. 

To achieve this, I construct a `DataArray` that maps the time steps of my dataset (`time` coordinate) to the start of each of such a season (values) and feed that to `groupby`.
```python
def _get_oct_may_index(ds):
    dd = ds.time.to_index().to_pydatetime()
    labels = []
    coords = []
    for d in dd:
        if d.month <= 5:
            refyear = d.year - 1
        elif d.month >= 10:
            refyear = d.year
        else:
            continue
        dref = datetime.datetime(refyear, 10, 1)
        labels.append(dref)
        coords.append(d)
    return xr.DataArray(labels, coords=dict(time=coords),
            dims='time', name='season_start')
```

I give it a custom `name='season_start'`, so I end up in [the last `else` of `GroupBy.__init__`](https://github.com/pydata/xarray/blob/v0.8.2/xarray/core/groupby.py#L223-L227), where my `DataArray` named `group` gets passed on to the function `unique_value_groups`. However, that function apparently expects a NumPy array, rather than a `DataArray`.

Please see [this ipynb](https://gist.github.com/j08lue/8ff16eacf638eafdde5b8439eba3d9b4) showing the error.

## Proposed solution

So it turns out this can easily be fixed by changing [line 226](https://github.com/pydata/xarray/blob/v0.8.2/xarray/core/groupby.py#L226) from

    unique_values, group_indices = unique_value_groups(group, sort=sort)

to 

    unique_values, group_indices = unique_value_groups(group.values, sort=sort)

Please see [this other ipynb](https://gist.github.com/j08lue/37c0d36af34482cf02a85422bfef6e19) where the result is as expected.

Now is this a bug or am I abusing the code somehow?

**Comment by shoyer** [comment]
This looks like a plausible fix, but what would be really helpful is a minimal, complete example that triggers the error. That should help clarify the issue and at the least, we will need that for a test case in your pull request.

**Comment by j08lue** [comment]
OK, here is the minimal example:

```python
import xarray as xr
import pandas as pd
import numpy as np

def test_groupy_datetime():
    """groupby with a DataArray of dtype datetime"""

    # create test data
    times = pd.date_range('2000-01-01', periods=4)
    data = np.random.rand(4)
    foo = xr.DataArray(data, coords=dict(time=times), dims='time')

    # create test index
    dd = times.to_datetime()
    labels = [dd[0]]*2 + [dd[2]]*2
    ind = xr.DataArray(labels, coords=dict(time=times), dims='time', name='reference_date')

    # group foo by ind
    foo.groupby(ind)
```

Making that, I found out that the problem only occurs when the DataArray used with `groupby` has **`dtype=datetime64[ns]`**.

The problem is that we effectively feed the DataArray to [`pd.factorize`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.factorize.html) and that goes well for most data types: Pandas checks with the function [`needs_i8_conversion`](https://github.com/pandas-dev/pandas/blob/v0.19.1/pandas/types/common.py#L248-L251) whether it can factorize the DataArray and decides YES for our `datetime64[ns]`. But then [in `pd.factorize`](https://github.com/pandas-dev/pandas/blob/v0.19.1/pandas/core/algorithms.py#L295-L307) it fails because it tries to access `DataArray.view` to convert to `int64`.

So there are three solutions to this:
1. Make Pandas' `pd.factorize` handle our `datetime` DataArrays better,
2. Add an attribute `.view` to DataArrays, or
3. Use the solution in the above PR, which means feeding only the NumPy `.values` to `pd.factorize`.

Which one do you prefer?

**Comment by shoyer** [comment]
Thanks for looking into this!

Based on how `factorize` works (with specialized handling for pandas dtypes), I think the most robust behavior would be to pass in a `pandas.Index`. For example, this will work better if someone uses a `pandas.PeriodIndex`. So I would suggest wrapping arrays with [`safe_cast_to_index`](https://github.com/pydata/xarray/blob/76726e58710122e302790a05ae092c30d41d3553/xarray/core/utils.py#L40) before passing them to `pd.factorize`.

I'm not sure if we want a `.view` attribute on DataArrays, but in any case it's not clear that would even fix the issue here -- pandas probably needs to a coerce to numpy arrays internally in factorize eventually anyways.

**Comment by j08lue** [comment]
Great, `safe_cast_to_index` works nicely (it passes my test). I added the change to the existing PR.

Do we need to add more groupby tests to make sure the solution is safe for other cases (e.g. other data types)?

**Comment by guziy** [comment]
Hi:

Here I have an example that have worked until recently...

https://github.com/guziy/PyNotebooks/blob/master/xarray/test_grouping.ipynb

Of course there is a resample function, but sometimes you want grouping...

Are there any plans to add a function as a key to the groupby function?

Cheers

**Comment by shoyer** [comment]
@guziy This should be fixed by #1133, which will be part of the next release.

**Comment by guziy** [comment]
Thanks, I'll try to use the github version.

Cheers
